diff --git a/src/utilities.cc b/src/utilities.cc
index 25c4b76..71f71e6 100644
--- a/src/utilities.cc
+++ b/src/utilities.cc
@@ -311,7 +311,7 @@ static void MyUserNameInitializer() {
     char buffer[1024] = {'\0'};
     uid_t uid = geteuid();
     int pwuid_res = getpwuid_r(uid, &pwd, buffer, sizeof(buffer), &result);
-    if (pwuid_res == 0) {
+    if (pwuid_res == 0 && result) {
       g_my_user_name = pwd.pw_name;
     } else {
       snprintf(buffer, sizeof(buffer), "uid%d", uid);
